File: cs-server/cs-server/Game.cs
using System;
using System.Collections;
using System.Collections.Generic;

namespace CS_Server
{
    class Game
    {
        private readonly Dictionary<int, Player> clients = new Dictionary<int, Player>();

        public void CreatePlayer(int clientId, int playerId, int hasBall)
        {
            Player client = new Player(clientId, playerId);
            client.giveBall(hasBall);
            clients.Add(playerId, client);
        }


        public List<int> getClientId(int clientId)
        {
            List<int> result = new List<int>();

            foreach (Player client in clients.Values)
            {
                if (client.getClientId == clientId)
                {
                    result.Add(client.getClientId);
                }
            }

            return result;

        }
        public List<int> getPlayer(int clientId)
        {
            List<int> result = new List<int>();
            foreach (Player client in clients.Values)
                if (client.getClientId == clientId)
                    result.Add(client.getPlayerId);
            return result;
        }

        public List<int> getPlayerIds()
        {
            List<int> result = new List<int>();
            foreach (Player client in clients.Values)
            {
                result.Add(client.getPlayerId);
            }
            return result;
        }

        public int getBall(int clientId, int playerId)
        {
            if (clients[playerId].getClientId != clientId)
                throw new Exception(
                    $"Client: {clientId} is not allowed to pass the ball for player {playerId}.");

            return clients[playerId].hasBall();
        }

        public void giveBall(int clientId, int fromPlayer, int toPlayer, int ball)
        {
            lock (clients)
            {
                if (clients[fromPlayer].getClientId != clientId)
                    throw new Exception(
                        $"Client: {clientId} is not allowed to pass the ball for player {fromPlayer}");
                if (clients[fromPlayer].hasBall() == 0)
                    throw new Exception(
                        $"You  are not allowed to pass the ball.");
                if (ball <= 0)
                    throw new Exception("");

                if (clients[fromPlayer].hasBall() == 1)
                {
                    clients[toPlayer].giveBall(0);
                }
            }
        }
    }
}

File: java-server/src/Game.java
import java.util.ArrayList;
import java.util.List;
import java.util.Map;
import java.util.TreeMap;

public class Game {
    private final Map<Integer, Player> clients = new TreeMap<>();

    public void createPlayer(int clientId, int playerId, int hasBall)
    {
        Player client = new Player(clientId, playerId);
        client.giveBall(hasBall);
        clients.put(playerId, client);
    }

    public List<Integer> getClientIds(int clientId) {
        List<Integer> result = new ArrayList<Integer>();

        for (Player client : clients.values())
            if (client.getClientId() == clientId)
                result.add(client.getClientId());

        return result;
    }

    public List<Integer> getPlayer(int clientId) {
        List<Integer> result = new ArrayList<Integer>();

        for (Player client : clients.values())
            if (client.getClientId() == clientId)
                result.add(client.getPlayerId());

        return result;
    }

    public List<Integer> getPlayerIds() {
        List<Integer> result = new ArrayList<Integer>();

        for (Player client : clients.values())
            result.add(client.getPlayerId());

        return result;
    }

    public int getBall(int clientId, int playerId) throws Exception {
        return clients.get(playerId).hasBall();
    }

    public void setBall(int player) throws Exception {
        synchronized (clients) {
            clients.get(player).giveBall(0);
        }
    }

    public void giveBall(int clientId, int fromPlayer, int toPlayer, int ball) throws Exception {
        synchronized (clients)
        {
            if (clients.get(fromPlayer).getClientId() != clientId)
                throw new Exception("Player: " + clientId + " is not allowed to pass the ball for " + fromPlayer + ".");
            if (clients.get(fromPlayer).hasBall() < ball)
                throw new Exception(
                        "Player: " + fromPlayer +  " does not have the ball.");
            if (ball <= 0)
                throw new Exception("The amount has to be a positive value.");
            clients.get(fromPlayer).giveBall(clients.get(fromPlayer).hasBall() - ball);
            clients.get(toPlayer).giveBall(clients.get(toPlayer).hasBall() + ball);
        }
    }
}
